### PR TITLE
[ESIMD] E2E test for slm_allocator API.

### PR DESCRIPTION
**Pull request by kbobrovs**
Complementary patch for https://github.com/intel/llvm/pull/7759.
Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: SYCL/ESIMD/slm_alloc.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+//         /    \
+//        /      v
+//       /       bar - uses slm_allocator(SLM_IN_BAR)
+//      v
+//     foo - uses slm_allocator(SLM_IN_FOO)
+// Test1 kernel SLM usage is SLM_IN_KERNEL + max(SLM_IN_BAR, SLM_IN_FOO).
+// SLM offset returned by the slm_allocator in foo and bar is the same and is
+// SLM_IN_KERNEL bytes.
+// Bar uses slightly bigger SLM frame than foo. It modifies values (adds 10) in
+// SLM resulting from foo, plus appends coule more '100's.
```
Comment (v-klochkov) [inline SYCL/ESIMD/slm_alloc.cpp]: 'coule'? looks like a misprint

File: SYCL/ESIMD/slm_alloc.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+constexpr int SLM_IN_KERNEL = LOCAL_SIZE * ELEM_SIZE;
+constexpr int SLM_IN_FOO = LOCAL_SIZE * ELEM_SIZE;
+constexpr int BAR_EXTRA_ELEMS = 2;
+constexpr int SLM_IN_BAR = SLM_IN_FOO + BAR_EXTRA_ELEMS * ELEM_SIZE;
+constexpr int SLM_TOTAL = SLM_IN_KERNEL + std::max(SLM_IN_FOO, SLM_IN_BAR);
+constexpr int BAR_MARKER1 = 10;
+constexpr int BAR_MARKER2 = 100;
+
+#ifdef FORCE_INLINE
+inline
```
Comment (v-klochkov) [inline SYCL/ESIMD/slm_alloc.cpp:51]: It is supposed to be FORCE_INLINE, but 'inline' alone may be treated as a recommendation to FE.
Should it be using 'always_inline' attribute, similar to this one: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/defines_elementary.hpp#L48

**Review by v-klochkov — APPROVED** [review]
Please see 2 non-blocking comments.

**Comment by v-klochkov** [comment]
@kbobrovs - would you please create one more test case. Such difficult/complex changes in sycl-post-link deserve it, IMO.
The test could be:
// format: function_name(how much allocates) -->calls--> callee_function(how much allocates)
kernel(10) --> a(100) -- >b(200)
kernel(10) --> b(200);

So, b() has 2 paths to it from kernel().

**Comment by kbobrovs** [comment]
> @kbobrovs - would you please create one more test case. Such difficult/complex changes in sycl-post-link deserve it, IMO. The test could be: // format: function_name(how much allocates) -->calls--> callee_function(how much allocates) kernel(10) --> a(100) -- >b(200) kernel(10) --> b(200);
> 
> So, b() has 2 paths to it from kernel().

Yes, more tests will be added, including unit tests

File: SYCL/ESIMD/slm_alloc.cpp (hunk 2 above)
Comment (kbobrovs) [reply SYCL/ESIMD/slm_alloc.cpp:51]: Makes sense, thanks.

File: SYCL/ESIMD/slm_alloc.cpp (hunk 1 above)
Comment (kbobrovs) [inline SYCL/ESIMD/slm_alloc.cpp]:
```suggestion
// SLM resulting from foo, plus appends couple more '100's.
```

File: SYCL/ESIMD/slm_alloc.cpp (hunk 2 above)
Comment (kbobrovs) [inline SYCL/ESIMD/slm_alloc.cpp:51]:
```suggestion
inline __attribute__((always_inline))
```

File: SYCL/ESIMD/slm_alloc.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    void
+    foo(int local_id) {
+  slm_allocator a(SLM_IN_FOO);
+  uint32_t slm_off = a.get_offset();
+  // write data chunk "Y":
+  slm_scalar_store(slm_off + local_id * ELEM_SIZE, (T)local_id);
+}
+
+#ifdef FORCE_INLINE
+inline
```
Comment (kbobrovs) [inline SYCL/ESIMD/slm_alloc.cpp:66]:
```suggestion
inline __attribute__((always_inline))
```

**Review by v-klochkov — APPROVED** [review]

File: SYCL/ESIMD/slm_alloc.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// RUN: %GPU_RUN_PLACEHOLDER %t.2.out
+
+// This is end-to-end test for the slm_allocator API used together with the
+// slm_init. The call graph is:
+//         Test1(kernel) - uses slm_init(SLM_IN_KERNEL)
+//         /    \
+//        /      v
+//       /       bar - uses slm_allocator(SLM_IN_BAR)
+//      v
+//     foo - uses slm_allocator(SLM_IN_FOO)
```
Comment (kbobrovs) [inline SYCL/ESIMD/slm_alloc.cpp]:
```suggestion
//      foo - uses slm_allocator(SLM_IN_FOO)
```

**Comment by v-klochkov** [comment]
The failed tests: accessor.cpp and UserDefinedReductions/user_defined_reductions_wg_size_larger_than_data_size.cpp are unrelated to the 3 tests being added by this PR. 
Merged.

**Comment by v-klochkov** [comment]
Thank you @kbobrovs for the changes and the test!

**Pull request merged**